### PR TITLE
fix!: in SmartWallet, if invitation is invalid, don't process offer

### LINKED ISSUE #9239: enhance smart wallet behavior when invitation creation fails
## What is the Problem Being Solved?

Currently smart-wallet lacks tests for failure to create invitation. This is a failure we want to handle to ensure we're handling invitations correctly in the smart wallet.

## Test Plan

Have test coverage for more edge cases.

## Upgrade Considerations

```[tasklist]
### Tasks
- [ ] https://github.com/Agoric/dapp-offer-up/pull/76
- [ ] https://github.com/Agoric/agoric-sdk/pull/9250
- [ ] https://github.com/Agoric/agoric-sdk/pull/9254
- [ ] https://github.com/Agoric/agoric-sdk/pull/9204
- [ ] https://github.com/Agoric/agoric-sdk/pull/9240
- [ ] https://github.com/Agoric/agoric-sdk/pull/9236
- [ ] https://github.com/Agoric/agoric-sdk/pull/9238
```

### PR DESCRIPTION
**Pull request by Chris-Hibbert**
refs: #9239

## Description

When SmartWallet gets a request to exercise an invitation, it shouldn't create an offer if the invitation is broken.

### Security Considerations

Creating an offer requires withdrawing funds, which have to be returned to the user. Better to check the invitation first and skip the rest.

### Scaling Considerations

None

### Documentation Considerations

None

### Testing Considerations

We added a test and a fix.

### Upgrade Considerations

We'd like to get this into the next available upgrade. There are no complex interactions with other updates.

### CODE REVIEW

File: packages/smart-wallet/src/smartWallet.js [hunk 1]
```diff
@@ -972,13 +972,14 @@ export const prepareSmartWallet = (baggage, shared) => {
 
             const invitation = invitationFromSpec(offerSpec.invitationSpec);
 
-            const [paymentKeywordRecord, invitationAmount] = await Promise.all([
+            const invitationAmount =
+              await E(invitationIssuer).getAmountOf(invitation);
```
Comment (dckc) [inline packages/smart-wallet/src/smartWallet.js:976]: Looks like we're reconsidering the design from...

 - https://github.com/Agoric/agoric-sdk/issues/7098

Is there anybody we should check with? The record only shows you, me, and @turadg as party to that issue and its fix.

File: packages/smart-wallet/test/test-invitation1.js [hunk 2]
```diff
@@ -33,6 +33,7 @@ const mockBootstrapPowers = async (
   log,
   spaceNames = ['installation', 'instance', 'issuer', 'brand'],
 ) => {
+  /** @type {import('@agoric/vat-data').Baggage} */
   const baggage = makeScalarMapStore('bootstrap');
```
Comment (dckc) [inline packages/smart-wallet/test/test-invitation1.js:37]: Does anybody know why the type of `baggage` isn't inferred from `makeScalarMapStore`?

File: packages/smart-wallet/test/test-invitation1.js [hunk 3]
```diff
@@ -151,6 +155,7 @@ test.serial('handle failure to create invitation', async t => {
   const invitationIssuer = powers.issuer.consume.Invitation;
   const address = 'agoric1234';
 
+  // @ts-expect-error It's a promise.
   const walletsStorage = E(chainStorage).makeChildNode('wallet');
```
Comment (dckc) [inline packages/smart-wallet/test/test-invitation1.js]:
```suggestion
  // @ts-expect-error Test setup ensures that chainStorage resolution is not undefined.
  const walletsStorage = E(chainStorage).makeChildNode('wallet');
```

another approach is something like:

```js
 * @param {BootstrapPowers & NonNullChainStorage} powers
 * @typedef {PromiseSpaceOf<{ chainStorage: StorageNode }>} NonNullChainStorage
```

**Review by dckc — APPROVED** [review]
a couple questions and a suggestion... not critical

File: packages/smart-wallet/test/test-invitation1.js (hunk 3 above)
Comment (turadg) [reply packages/smart-wallet/test/test-invitation1.js]: I think either of the above is good for this PR. You could also reference,
- https://github.com/Agoric/agoric-sdk/issues/8247

File: packages/smart-wallet/test/test-invitation1.js (hunk 2 above)
Comment (turadg) [reply packages/smart-wallet/test/test-invitation1.js:37]: `Baggage` is `MapStore<string, any>` but `makeScaleMapStore` returns `MapStore<unknown, unknown>`.

It can take `keyShape` and `valueShape` in options, so it could eventually infer those two type parameters, but that would depend on the work to infer types from shapes.

File: packages/smart-wallet/src/smartWallet.js [hunk 4]
```diff
@@ -972,13 +972,14 @@ export const prepareSmartWallet = (baggage, shared) => {
...
+
+            const paymentKeywordRecord =
```
Comment (turadg) [inline packages/smart-wallet/src/smartWallet.js]: This is mutative, so no longer simply "prepare". It belongs under "2. Begin executing offer". Consider also naming the object like `withdrawnPayments` to it's more clear how it relates to the lifecycle of the request.

**Review by turadg — CHANGES_REQUESTED** [review]

**Comment by dckc** [comment]
Why the `!` breaking change marker?

File: packages/smart-wallet/test/test-invitation1.js (hunk 2 above)
Comment (dckc) [reply packages/smart-wallet/test/test-invitation1.js:37]: > that would depend on the work to infer types from shapes

adding a link to track motivations:

 - https://github.com/Agoric/agoric-sdk/issues/6160

File: packages/smart-wallet/src/smartWallet.js (hunk 1 above)
Comment (Chris-Hibbert) [reply packages/smart-wallet/src/smartWallet.js:976]: If I was part of that discussion, I've forgotten it.

Thanks for the off-line discussion. I'm comfortable that given what we know now, this order is better.

File: packages/smart-wallet/src/smartWallet.js (hunk 4 above)
Comment (Chris-Hibbert) [reply packages/smart-wallet/src/smartWallet.js]: done.

File: packages/smart-wallet/test/test-invitation1.js (hunk 3 above)
Comment (Chris-Hibbert) [reply packages/smart-wallet/test/test-invitation1.js]: added

**Comment by Chris-Hibbert** [comment]
> Why the ! breaking change marker?

I think of *!* as "incompatible behavior change", but if its really "breaking change", I'll drop it.

**Review by turadg — APPROVED** [review]
